Geolocator tests refactored after talking to Maciek

siejkowski opened the pull request.

literator commented on RxSwiftWorkshop/Maps/Model/GeolocatorSpec.swift:
```diff
@@ -34,47 +33,89 @@ class GeolocatorSpec: QuickSpec {
-                    withIdentifier: FakeSystemGeolocator.Identifiers.geocodeAddressString, arguments: [address]
-                )
+                setup {
+                    sut?.geolocate(address: address).subscribe().disposed(by: disposeBag!)
```
I think, if we IUO disposeBag, we can do it above (`var disposeBag: DisposeBag!`). The result is the same and you don't have to remember about `!` each time, and it reads nicer.

siejkowski replied:
You won't have to remember that – compiler will tell you. Also, the reason why I prefer avoiding the IUOs is that they are stating the possibility of crash only in the variable declaration, not in the actual place of crashing, which is the variable use. Dropping IUO and bringing force unwrapping to the place of use brings the expression of "here we could crash" back to the actual place of crashing. If it's not a deal breaker for you,  I'm gonna keep wringing this way.

literator replied:
I like the idea of a helper (like your `unwrap`) that actually allows to print something more when unwrap crashes. I agree that _simple_ crash says only that _Unwrapping optional nil failed_ and that's it.

literator commented on RxSwiftWorkshop/Maps/Model/GeolocatorSpec.swift:
```diff
@@ -34,47 +33,89 @@ class GeolocatorSpec: QuickSpec {
+                    sut?.geolocate(address: address).subscribe().disposed(by: disposeBag!)
+                }
+                it("should pass geolocation request to its SystemGeolocator") {
+                    systemGeolocator!.verifyCall(
```
I'd follow that `!` rule everywhere. (see previous comment)

Or is that something I miss, and those two cases are not identical?

siejkowski replied:
Just as before.

literator replied:
Yeah

literator commented on RxSwiftWorkshop/Maps/Model/GeolocatorSpec.swift:
```diff
@@ -34,47 +33,89 @@ class GeolocatorSpec: QuickSpec {
-                expect(result?.coordinate).to(equal(location.coordinate))
+                setup {
+                    systemGeolocator!.fixedLocation = location
+                    sut!.geolocate(address: "Valencia Airport")
```
I'd use RxText here with testable observers and so on:
```swift
var scheduler: TestScheduler! // = TestScheduler(initialClock: 0)
(...)
var testableObserver: TestableObserver<CLLocation>!
(...)
systemGeolocator.fixedLocationObservable = scheduler.createColdObservable([next(0, location), completed(0)])
testableObserver = scheduler.start {
  sut.geolocate(address: "Valencia Airport")
}

(...)

it("...") {
  expect(testableObserver.events[0].value.element).to(equal(expectedLocation))
}
```

However, I was doing Rx tests like you did here for some time as well. And here it is does its job.
But RxTest is safer in general, when asynchrony is being involved and you really need to control virtual time. Thus, for the sake if consistency I'd use that approach even if there's no asynchrony and there's no good reason not to.

What do you think?

siejkowski replied:
I'm still struggling with this one. `SystemGeolocator` has no reactive API nor anything to do with reactiveness at all, and therefore it's strange for `FakeSystemGeolocator` to have information wrapped into observables instead of fixtures. The asynchronicity, however, might take place (because of `completionBlock` in `SystemGeolocator`). The way we deal with that is the implementation of `FakeSystemGeolocator`. Whether `FakeSystemGeolocator` takes an observable or a fixture, it's in its implementation details to pass the location in an synchronous way. I don't really see how the scheduler changes that. Could you elaborate further?

literator replied:
I think, I made a mistake for the second time. You're right – system geolocator doesn't expose reactive API. So the `testableObservable` doesn't make much sense, which is a pity, because that would be nice.

In this case you have capture next result, indeed. I don't like it very much, but there's nothing you can actually do. 

But you might be using `RxBlocking`, which allows you to write:

```swift
let result = sut.geolocate(address: "Valencia Airport").toBlocking().first
```

It makes asynchronious more synchronious in a, not very nice, way.

literator replied:
Och, and in this case `TestScheduler` doesn't help much.
It helps you manipulating virtual time, but only when you have cold/hot observables created by it.

siejkowski commented:
Commenting on the review summary.

> frankly I had to focus much more when reading, because I'm used to common beforeEach, afterEach, it, describe, context

I can definitely relate, since now I also feel I need focus more when I read tests written in a style of breaking the test case into separate blocks using `context` and `beforeEach` instead of keeping all the logic particular to the test case in `it`.

Concerning the new DSL words, I've dropped `describes`. All it does is takes into consideration that it's just after `spec`, which makes for a nice English sentence. One of the goals of BDD style of tests is to provide a easily readable specs, so in principle using `describes` is ok, but in practice it's not worth bringing additional symbol to the vocabulary.

I've also dropped `condition`, since it has a somewhat nuanced semantic difference from `context`. While I think it describes better the role that usually `context` plays – explain in what state the SUT and environment around it are for the test to make sense – it's also not worth bringing additional symbol.

The only word that I feel is actually worth it is `setup`, and I'm gonna leave setup in. Using `beforeEach` for separating `when` from `then` is a solution that plays nicely with the functional side of Quick (the order of execution, the scope availability). It keeps, however, little relation to the meaning of the function, which says "this is the part of setup that is common for all the tests in the scope", not "this is the setup for one test".

All in all, I'm glad I've come this short journey through finding a more descriptive names for the blocks that Quick provides, because it helped me understand better what is their role in the scene of this particular way of writing BDD tests.

At the end, just a quick response to two last points:

> 1. All those keywords like describe and context are sometimes misused or they are points of discussion or misunderstanding. If you introduce two more, then you have more difficult DSL to use, less clear, and more occasions for, maybe, needless discussions.

I somewhat disagree. If you use the DSL in a way that it was not designed to be used, than providing more words is actually beneficial to the expressiveness of the DSL. Quick does not shy from bringing separate words for what is functionally identical but plays a semantically different role – see `describe` and `context`. There's no need for us to shying from in neither. And if it brings the discussions, well, I think they should not be seen as `needless`. They are a part of a process of understanding the domain.

> 2. All in all, Quick keywords are in the vocabulary of every iOS dev. If you pass the project to someone else they might have troubles understating what's going on. But I don't feel it's strong arguments.

First of all, every iOS dev? While BDD style tests are popular, assuming that they are popular to the point of coming into vocabulary of every iOS dev is a huge overstatement. Even the Apple frameworks are not always part of common iOS dev vocabulary, let alone third-party test framework. 

Apart from that, this is an excellent point and a reason why in an app that actually will be maintained I'd be more conscious about bringing the additional identifiers. But, for one, we are not in this situation here, and for two, even then bringing team-agreed and team-driven improvement to DSL is not a bad idea. While the maintainability of our code is one of the main principles, it's not an only principle. Otherwise, how could we use a framework like RxSwift, which brings a mental model that is new and alien for many developers?

literator commented:
>I can definitely relate, since now I also feel I need focus more when I read tests written in a style of breaking the test case into separate blocks using context and beforeEach instead of keeping all the logic particular to the test case in it.

That's probably a matter of a habit.

Regarding introducing new DSL words, I'm not agains it in general. However I'd be very careful. 
When I had troubles expressing a behavior I wanted to test, usually it was my fault or a flaw in my thinking, rather than lack of proper DSL words. They somewhat _force_ me to structure my 'spec' thinking.

> And if it brings the discussions, well, I think they should not be seen as needless. They are a part of a process of understanding the domain.

Agreed. That's maybe because I'm recently a bit tired of overthinking and/or over engineering some concepts all the time around.

> First of all, every iOS dev?

Of course, I was too positive here. I meant "every iOS dev that writes tests".

> But, for one, we are not in this situation here, and for two, even then bringing team-agreed and team-driven improvement to DSL is not a bad idea. While the maintainability of our code is one of the main principles, it's not an only principle. Otherwise, how could we use a framework like RxSwift, which brings a mental model that is new and alien for many developers?

Fair point. That's why we have this discussion.
